### DISCUSSION TITLE
Allow disabling more features.

### DISCUSSION THREAD

**Discussion by fxzxmicah** [storaged-project/udisks#1314] [Ideas]
For example, raid and crypto.

**Comment by tbzatek** [comment]
Unfortunately the API is not ready for this and consumers expect the functionality to be always present.

**Reply by fxzxmicah** [reply]
I mean that these features can be disabled during build, but of course they can be set to be included by default.

**Reply by tbzatek** [reply]
I understand what you mean, but as I said above, this will break existing consumers when part of an API suddenly becomes unavailable.

**Reply by fxzxmicah** [reply]
When consumers use default build options, won't they lose anything? I'm referring to advanced users who use the build options that suit their own needs.

**Reply by tbzatek** [reply]
UDisks primarily provides public API which comes with certain guarantees that consumers built their code upon, there's no concept of advanced or basic user. UDisks is a middle layer, not an end user application.

**Reply by fxzxmicah** [reply]
I still haven't figured out what you mean. Why does it break existing consumers when I build my own UDisks solely for my own computer?
Or will separating certain features disrupt the functionality of UDisks? I see that btrfs is already a feature that can be disabled.

**Reply by vojtechtrefny** [reply]
> Why does it break existing consumers when I build my own UDisks solely for my own computer?

You are free to change anything you want for *your own computer*, but if *we* offer such option then we need to support it and make sure the public API we provide is stable and the behaviour of udisks isn't changed, which we cannot do when providing such option. Unfortunately saying "we provide this option, but don't use it if you are not the sole end user" doesn't really work.

>  I see that btrfs is already a feature that can be disabled.

That's the difference between a plugin and a core feature. Would it better if MD support was a plugin? Maybe, but it isn't. We didn't make this choice. Could we make it a plugin? Maybe in a new major release with a new API version, but we definitely don't have capacity for doing that and maintaining two versions of udisks.

**Reply by fxzxmicah** [reply]
Okay, I understand your explanation. Sorry to bother you.

**Reply by tbzatek** [reply]
There are several aspects to making something suddenly optional, as @vojtechtrefny explained. Even if you build something on your own machine doesn't guarantee that any UDisks consumer on your machine will work as expected. We've seen number of weird errors on patched UDisks.

Plugins are entirely optional and have been designed for that. There's a mechanism for an application to check whether the plugin is available and act upon that. But the core API that was not designed with this in mind might have unforeseen consequences. It's always about the balance to define what are the core features and what can be made optional, a decision that is supposed to be made on every new addition.

One good example is the ATA SMART functionality that we've made recently optional - simply because the API was ready for that from the beginning.

**Reply by fxzxmicah** [reply]
Okay, I have understood the current situation. Although I hope to implement this feature, I understand that it is difficult to achieve. Thank you for your answer.
